src/server/pfs/server/driver.go

```diff
@@ -1997,22 +1975,45 @@ func (d *driver) subscribeCommit(pachClient *client.APIClient, repo *pfs.Repo, b
 		case watch.EventError:
 			return event.Err
 		case watch.EventPut:
-			if err := event.Unmarshal(&branchName, branchInfo); err != nil {
+			if err := event.Unmarshal(&commitID, commitInfo); err != nil {
 				return fmt.Errorf("Unmarshal: %v", err)
 			}
-			if branchInfo.Head == nil {
-				continue // put event == new branch was created. No commits yet though
+			if commitInfo == nil {
+				return fmt.Errorf("Commit info is empty for id: %v", commitID)
+			}
+
+			// if provenance is provided, ensure that the returned commits have the commit in their provenance
+			if prov != nil {
+				valid := false
+				for _, cProv := range commitInfo.Provenance {
+					if cProv.Commit.Repo.Name == prov.Commit.Repo.Name &&
+						cProv.Commit.ID == prov.Commit.ID &&
+						cProv.Branch.Name == prov.Branch.Name {
+						valid = true
+					}
+				}
+				if !valid {
+					continue
+				}
 			}
 
-			// TODO we check the branchName because right now WatchOne, like all
-			// collection watch commands, returns all events matching a given prefix,
-			// which means we'll get back events associated with `master-v1` if we're
-			// watching `master`.  Once this is changed we should remove the
-			// comparison between branchName and branch.
+			if commitInfo.Branch != nil {
+				// For now, we don't want stats branches to have jobs triggered on them
+				// and this is the simplest way to achieve that. Once we have labels,
+				// we'll use those instead for a more principled approach.
+				if commitInfo.Branch.Name == "stats" {
+					continue
+				}
```

Comment:
It might make more sense to filter out commits from the `"stats"` branch in the worker master, it would make this code cleaner. And it seems like it's a better separation of concerns, since ideally pfs wouldn't need to know anything about the `"stats"`.

Reply:
Thoughts on this?

Reply:
I tried moving it there, but that didn't work. I didn't want to spend to much time on it because the right solution is just to use labels, which will be pretty easy to finally implement once this is in.

Reply:
KK, sounds good to me, wanna add a TODO here so we have a shot at remembering to remove it once labels are in? This does mean that it's impossible to to get stats commits in `SubscribeCommit` right?

Reply:
Yeah, stats can't get in it in this version. I'll get labels done right after this one.